internal/compression/zstd.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compression
+
+import (
+	"io"
+
+	"connectrpc.com/connect"
+	"github.com/klauspost/compress/zstd"
+)
+
+// ZstdDecompressor is a thin wrapper around a zstd Decoder.
+type ZstdDecompressor struct {
+	decoder *zstd.Decoder
+}
+
+func (c *ZstdDecompressor) Read(bytes []byte) (int, error) {
+	return c.decoder.Read(bytes)
+}
+func (c *ZstdDecompressor) Reset(rdr io.Reader) error {
+	return c.decoder.Reset(rdr)
+}
+func (c *ZstdDecompressor) Close() error {
+	c.decoder.Close()
+	return nil
+}
+
+// NewZstdDecompressor returns a new Zstd Decompressor.
+func NewZstdDecompressor() connect.Decompressor {
+	d, _ := zstd.NewReader(nil)
+	return &ZstdDecompressor{
+		decoder: d,
+	}
+}
+
+// ZstdCompressor is a thin wrapper around a zstd Encoder.
+type ZstdCompressor struct {
+	*zstd.Encoder
+}
+
+// NewZstdCompressor returns a new Zstd Compressor.
+func NewZstdCompressor() connect.Compressor {
+	w, _ := zstd.NewWriter(nil)
```

Comment:
We have to swallow this error here because the `connect.WithAcceptCompression` method expects a function that just returns a `connect.Compressor` (not a compressor and error) and the same goes for the decompressor.

Reply:
This is bound to cause a panic if an error ever does occur. Eating errors is almost never acceptable -- often better to panic so that at least the code fails fast with a good error message instead of possibly panic'ing later with an inexplicable nil-pointer dereference.

In this case, instead of panic'ing or ignoring, there is another option (which should also be used above in `NewZstdDecompressor`): return a sentinel, like an `errorCompressor` or `errorDecompressor`. Those types can have a single `err error` field and _always_ return that error from every method. That way you can see the error that happened here, but it will be delayed to the point of first actual use of the compressor/decompressor.

Reply:
Awesome idea, thank you. Updated.